core/src/main/java/com/linecorp/armeria/internal/ArmeriaHttpUtil.java

```diff
@@ -489,8 +515,19 @@ public static long parseCacheControlSeconds(@Nullable String value) {
     /**
      * Converts the specified Netty HTTP/2 into Armeria HTTP/2 headers.
      */
-    public static HttpHeaders toArmeria(Http2Headers headers, boolean endOfStream) {
-        final HttpHeaders converted = new DefaultHttpHeaders(false, headers.size(), endOfStream);
+    public static HttpHeaders toArmeria(Http2Headers headers, boolean request, boolean endOfStream) {
+        final HttpHeadersBuilder converted;
+        if (request) {
+            converted = headers.contains(HttpHeaderNames.METHOD) ? RequestHeaders.builder()
```

Comment:
Might have missed it, but I don't think there are new unit tests for these conditionals. Should we add?

Reply:
Added a test case. :smile: